### PR TITLE
Provide a way to customize pipeline behavior for client side

### LINKED ISSUE #3907: Provide a way to limit traffic for client side
**Motivation**

Externally, users may want to limit the amount of ingress/egress traffic in bytes.

Internally, `armeria` developers have no real way to test/simulate edge cases when we have slow connections.
For instance, there is no integration test to check `WriteTimeoutException` behavior in our code base.

I propose that we provide an option where `ChannelTrafficShapingHandler` is added to the client pipeline conditionally.

i.e.

`ChannelTrafficShapingHandler.java`
```
public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, SocketAddress localAddress,
                        ChannelPromise promise) throws Exception {
...
    final ChannelPipeline p = ch.pipeline();
    p.addLast(new FlushConsolidationHandler());
    p.addLast(ReadSuppressingAndChannelDeactivatingHandler.INSTANCE);
    if (clientFactory.writeLimit > 0 || clientFactory.readLimit > 0) {
        p.addLast(new ChannelTrafficShapingHandler(clientFactory.writeLimit, clientFactory.readLimit));
    }
...
}
```

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

- Currently, there is no easy way to create integration tests which simulate certain behaviors (e.g. `WriteTimeoutException`) 
    - #4259 
- Users may also want to customize behavior like throttling, fault injection, etc. for clients.

Modifications:

- Add `CHANNEL_PIPELINE_CUSTOMIZER` to `ClientFactoryOptions`, which is a consumer on `ChannelPipeline`
    - I'm a little skeptical of adding these methods to `ClientFactoryBuilder`, `Flags` until more users want this feature.
- Call `HttpClientFactory#channelPipelineCustomizer` right before `channel#connect`, giving users to customize the pipeline at the last minute.
- Add a unit test which simulates write timeout behavior.
- Rename `invokeProxyConnectFailed` to `maybeHandleProxyFailure` to better represent what the method is doing.

Result:

- Closes #3907

### CODE REVIEW

**Comment by jrhee17** [comment]
note: the CI will pass once #4259 is merged

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 1]
```diff
@@ -263,6 +263,26 @@ private static long clampedDefaultMaxClientConnectionAge() {
                 return newValue.option().newValue(builder.build());
             });
 
+    /**
+     *  The number of bytes per sec allowed to be read for each channel.
+     *  A value of {@code 0} signifies no limit is applied.
+     */
+    public static final ClientFactoryOption<Long> WRITE_BYTES_PER_SEC_LIMIT =
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:270]: Naming: We usually start an option name with `maxXXX` and `minXX` when the option limits something.
Should we change the naming style to `MAX_WRITE_BYTES_PER_SEC`?

Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:270]: Add `@Unstable`?

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 2]
```diff
@@ -263,6 +263,26 @@ private static long clampedDefaultMaxClientConnectionAge() {
...
+            ClientFactoryOption.define("WRITE_BYTES_PER_SEC_LIMIT", 0L, v -> {
+                checkArgument(v >= 0);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:272]:
```suggestion
                checkArgument(v >= 0, "WRITE_BYTES_PER_SEC_LIMIT %s (expected: >= 0)", v);
```

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 3]
```diff
@@ -526,4 +546,20 @@ public Http1HeaderNaming http1HeaderNaming() {
     public boolean tlsAllowUnsafeCiphers() {
         return get(TLS_ALLOW_UNSAFE_CIPHERS);
     }
+
+    /**
+     * Returns the number of bytes per sec allowed to be written for each channel.
+     * A value of {@code 0} signifies no limit is applied.
+     */
+    public long writeBytesPerSec() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:554]:
```suggestion
    @UnstableApi
    public long maxWriteBytesPerSec() {
```

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 4]
```diff
@@ -526,4 +546,20 @@ public Http1HeaderNaming http1HeaderNaming() {
...
+        return get(WRITE_BYTES_PER_SEC_LIMIT);
+    }
+
+    /**
+     * Returns the number of bytes per sec allowed to be read for each channel.
+     * A value of {@code 0} signifies no limit is applied.
+     */
+    public long readBytesPerSec() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:562]:
```suggestion
    @UnstableApi
    public long maxReadBytesPerSec() {
```

File: core/src/test/java/com/linecorp/armeria/client/WriteTimeoutTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service(Route.ofCatchAll(), (ctx, req) -> HttpResponse.of(200));
+        }
+
+        @Override
+        protected void configureWebClient(WebClientBuilder webClientBuilder) throws Exception {
+            webClientBuilder
+                    .writeTimeoutMillis(1)
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/WriteTimeoutTest.java:52]: It is hard to test millisecond precision in CI environments.
How about setting the write timeout to 1 second and increasing the header size?

**Review by ikhoon — COMMENTED** [review]
Looks like useful options for testing!

**Comment by jrhee17** [comment]
> That being said, how about allowing a user to customize the Netty channel pipeline when a new channel is initialized? This way, a user can inject whatever behavior into the traffic or even implement an alternative protocol using it.

I agree with this point. At first I was a little skeptical of giving the user "too much power" so that users can break `armeria` behavior, but I think it's fine as long as it is documented clearly that 1) usage of this option is for advanced users only 2) it is recommended to consult with maintainers before using this feature seriously (so that `armeria` can provide more user-friendly APIs)

I've added a customizer for client-side only for now, let me create a separate issue for server-side 🙏

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 6]
```diff
@@ -263,6 +266,19 @@ private static long clampedDefaultMaxClientConnectionAge() {
                 return newValue.option().newValue(builder.build());
             });
 
+    /**
+     * This consumer lets users customize the behavior of Netty's {@link ChannelPipeline}.
+     * This customizer is run right before {@link ChannelPipeline#connect(SocketAddress)}
+     * is invoked by armeria. This customizer is no-op by default.
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:272]: 🦅👁
```suggestion
     * is invoked by Armeria. This customizer is no-op by default.
```

**Review by ikhoon — APPROVED** [review]
Looks better!

**Review by minwoox — APPROVED** [review]
What a brilliant feature! 😄

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 7]
```diff
@@ -526,4 +542,13 @@ public Http1HeaderNaming http1HeaderNaming() {
     public boolean tlsAllowUnsafeCiphers() {
         return get(TLS_ALLOW_UNSAFE_CIPHERS);
     }
+
+    /**
+     * Returns the {@link ChannelPipeline} customizer set. This consumer is no-op
+     * by default.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:548]: Could you copy the description in `#CHANNEL_PIPELINE_CUSTOMIZER`?

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 8]
```diff
@@ -263,6 +266,19 @@ private static long clampedDefaultMaxClientConnectionAge() {
                 return newValue.option().newValue(builder.build());
             });
 
+    /**
+     * This consumer lets users customize the behavior of Netty's {@link ChannelPipeline}.
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java]: Thanks, done!

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 9]
```diff
@@ -263,6 +266,19 @@ private static long clampedDefaultMaxClientConnectionAge() {
...
+     * This customizer is run right before {@link ChannelPipeline#connect(SocketAddress)}
+     * is invoked by Armeria. This customizer is no-op by default.
+     *
+     * <p>Note that this customizer usage of this customizer is an advanced
+     * feature, and may produce unintentional side effects. It is encouraged to
+     * consult with maintainers before using this feature seriously.
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java]: Done!

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java (hunk 7 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java]: 👍

**Comment by minwoox** [comment]
Thanks a lot, @jrhee17! 😄
